### ISSUE TITLE
Contacts color should match in group messages

### ISSUE THREAD

**Issue by Captainjman2** [klinker-apps/messenger-issues#35]
**Device with issue:** Nexus 6P running Android 7.0
**App version:** 1.1.209

**Additional information:**

Right now all messages in the group chat are the same color and do not match the sending contacts set color.

**Possible solution:**

The contacts color used in the app should match the color of the received message from that contact in a group message.

**Comment by klinker41** [comment]
Its something I've done in the past but it is a little slow so I took it out. Thanks for the request though, I will see about rethinking it in the future.

**Comment by klinker41** [comment]
An implementation idea: 

When a group message is received, look up the sender's contact picture. If he has one, save the primary color from it with the message in the database (there is already a `color` column here that I'm not using, built it in specifically for this purpose). 

When displaying the message list, if a message has a color then just use that. If it doesn't, look at the contact name and find out if we already have a color generated for this person (save these colors to a `Map` inside the adapter so that they can be reused). If we do, then use that one. If we don't, generate one and write the primary color to a `Map<String, Integer>` with the contact name as the key.

One potential problem with this is that contact's colors will change if you leave the conversation and go back in and we weren't able to find a primary color for their contact picture. No way around this as I don't want to have to update the message in the database, too slow and hard to keep track of as new messages come in.

That will work for the android app, will have to think about the web app more in depth.

**Comment by klinker41** [comment]
To implement this the right way will take a lot of work, quite a few things are going to need to get done:

* [ ] Create a table on the device to store contact information. `contacts` table should include 6 columns, `name`, `number`, and `color`, `color_dark`, `color_accent`, `color_light`.
* [ ] When a new conversation is created, we should check this table to see if the contact already exists. If so, use the colors already set for this person. If not, create a new contact, generate colors (either through their photo or random material scheme), and use those colors for the conversation.
* [ ] When a group message is received, check if the table already exists for the contact. If so, save the `message.color` as the `contact.color`. If not, generate colors from their contact picture or random material scheme, save that contact and use those colors for the message.
* [ ] Add a table on the server to store this information.

This will keep a contact's colors consistent throughout the app, even if they don't have a contact picture which before would have been impossible.

Also, this will create a good stepping stone for https://github.com/klinker-apps/messenger-issues/issues/16. By getting the tables and interaction set up between devices. We would just need to implement uploading all contacts on the initial sign up then when we upload messages and conversations (in the `ApiUploadService`).

**Comment by klinker24** [comment]
Backend configuration for contacts model: 

https://github.com/klinker41/messenger-server/commit/f7959a996bff7e965aefd62e95a879a99f96f3b6
https://github.com/klinker41/messenger-server/commit/6717efa335e5e777b8ed720ad1a6e0d87db08920
https://github.com/klinker41/messenger-server/commit/b89de10054d9d8460e265f2f90350b1512c7f2a0

**Comment by klinker24** [comment]
Backend integration for Android side of things:

- [ ] create contacts table on android device
- [ ] create contacts api through `retrofit`
- [ ] make datasource CRUD methods
- [ ] integrate datasource CRUD methods with `retrofit` apis
- [ ] implement firebase receivers for "added_contact", "updated_contact", and "removed_contact"

This list does not include the actual implementation of the datasource methods that will be needed to sync the data between devices and backend.

**Comment by klinker24** [comment]
Retrofit APIs: https://github.com/klinker41/messenger/commit/250d1ffa702948f0ab1cb1d7bbc9ecf2cfc50181 and https://github.com/klinker41/messenger/commit/8a3c3adedaea31c8226659a27216a3d6d6e9956d
Contacts table: https://github.com/klinker41/messenger/commit/2ed197b59e3ea2f212d17978756a18a808e49592
CRUD methods: https://github.com/klinker41/messenger/commit/0c342275aadf5f0d424ddaf219ab2014d0be11ac
Push notifications: https://github.com/klinker41/messenger/commit/b97a62e3167a8d09df5ce2d0623918c3e58bb876

**Comment by klinker24** [comment]
Done:

- Android fills the contacts database on the initial install of the app: https://github.com/klinker41/messenger/commit/315f85b1443a9e21a4b8b2ed16d4d8c1e9271246
- Android uploads contact data encrypted: https://github.com/klinker41/messenger/commit/acc7a9c0e23e3c7d9b2f5527b74f98255d6f7811
- Show different colors for messages received in group conversations: https://github.com/klinker41/messenger/commit/891b79b2f6bc5d40f8a012f10e720bb97d67a8c9

I didn't follow quite the format that you suggested @klinker41 . More of what you said the first time, plus the database connection to make the colors persist across sessions.

If a group conversation is pulled up, it will grab the people and assign them a color, either from our database or a randomly generated one if the user is not a saved contact. If the user is not a saved contact, save them to the database with the color scheme.

I never update message colors in the database with this approach, I feel as though that just isn't a good way to do it, would put strain on our server every single new message. Offload that strain to the user's device instead. I also never update the conversation color based on the users that are involved. 

So the conversation retains the color that is initially generated for it, or uses the color of that conversation's preference. This means that the conversation color doesn't correspond to the contact color, even on solo conversations with that user, but I think that that is just fine, user won't notice a difference. This also means that you cannot change a user's color in a group conversation.

I am not going to make these changes on the web. We can open up a new story for that in the future if it is really necessary, but I do not think that it is.

@klinker41 reopen if you want this done a different way. If you think that this is fine, we can actually remove the color_dark, color_light, color_accent from the user table on android and on the backend, since we ever only use the normal color column.

**Comment by klinker41** [comment]
I think that is fine, looks good. However, we can't remove the dark, light and accent colors. We should be using those when a new conversation gets received to initialize it's colors.

**Comment by klinker24** [comment]
I changed some things tonight. Contact colors now sync with conversation colors for improved visibility.

**Comment by klinker41** [comment]
One fix for this https://github.com/klinker41/messenger-server/commit/bb7440f19d7e9c56f397752640299a9f147c380a

We need to delete the contacts table for an account before we can delete the account because of the foreign keys. This was breaking account deletion and in turn causing problems with the login/signup process for users who had removed their account in the app. Fix should be deployed and working now.
